incubator/hnc/internal/pkg/selectors/selectors.go

```diff
@@ -0,0 +1,70 @@
+// Package selectors contains exceptions related utilities, including check the validity of selector,
+// treeSelector, and noneSelector, and parsing these three selectors.
+package selectors
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+
+	api "sigs.k8s.io/multi-tenancy/incubator/hnc/api/v1alpha2"
+)
+
+func GetSelectorAnnotation(inst *unstructured.Unstructured) (string, bool) {
+	annot := inst.GetAnnotations()
+	str, ok := annot[api.AnnotationSelector]
+	return str, ok
+}
+
+// GetTreeSelector is similar to a regular selector, except that it adds the LabelTreeDepthSuffix to every string
+// To transform a tree selector into a regular label selector, we follow these steps:
+// 1. get the treeSelector annotation if it exists
+// 2. convert the annotation string to a slice of strings seperated by comma, because user is allowed to put multiple selectors
+// 3. append the LabelTreeDepthSuffix to each of the treeSelector string
+// 4. combine them into a single string connected by comma
+func GetTreeSelector(log logr.Logger, inst *unstructured.Unstructured) (labels.Selector, error) {
+	annot := inst.GetAnnotations()
+	treeSelectorStr, ok := annot[api.AnnotationTreeSelector]
+	if !ok {
+		return nil, errors.New("TreeSelector not found.")
+	}
+	strs := strings.Split(treeSelectorStr, ",")
+	selectorStr := ""
+	for i, str := range strs {
+		selectorStr = selectorStr + str + api.LabelTreeDepthSuffix
+		if i < len(strs)-1 {
+			selectorStr = selectorStr + ", "
+		}
+	}
+	return getSelectorFromString(log, selectorStr)
+}
+
+// GetSelector returns the selector on a given object if it exists
+func GetSelector(log logr.Logger, inst *unstructured.Unstructured) (labels.Selector, error) {
+	selectorStr, ok := GetSelectorAnnotation(inst)
+	if !ok {
+		return nil, errors.New("Selector not found.")
+	}
+	return getSelectorFromString(log, selectorStr)
+}
+
+// getSelectorFromString converts the given string to a selector
+// Note: any invalid Selector value will cause this object not propagating to any child namespace
+func getSelectorFromString(log logr.Logger, str string) (labels.Selector, error) {
+	labelSelector, err := v1.ParseToLabelSelector(str)
+	// TODO: surface the error messages here (https://github.com/kubernetes-sigs/multi-tenancy/issues/1165)
+	if err != nil {
+		log.Error(err, "Could not parse selector annotation to labelSelector")
```

Comment:
As a general rule, errors should _either_ be returned _or_ logged. The reason is that you never want to see the same error cause lots of different log messages, which is what will happen if every function both logs and returns it.

What _is_ useful is [wrapping an error with more information](https://blog.golang.org/go1.13-errors). E.g., let's assume that both `v1.ParseToLabelSelector` and `v1.LabelSelectorAsSelector` both return errors that clearly identify which function returned the error (which is a good assumption). In that case, this function (`getSelectorFromString`) doesn't have any useful information to add. But the _callers_ do - they know if the error was caused by a `/selector` or `/treeSelector`. So, for example, in `GetSelector`, you might do something like this:

```
selector, err := getSelectorFromString(selectorStr)
if err != nil {
  return nil, fmt.Errorf("while parsing %q: %w", api.AnnotationSelector, err)
}
return selector, nil
```

In the validator, we can then pass this message back to the user (I _think_ the message will naturally show the object being rejected, so we don't need to add that information here). In the reconciler, we _don't_ want to return the error to controller-runtime, since it's a _user_ error and we don't want to re-reconcile the object. So that's a good place to log the error - and the logger already knows the name of the affected object, so we don't need to add it again there either.

Does this make sense?

Reply:
I see! Done